src/System.Linq/src/System/Linq/Enumerable.cs

```diff
@@ -676,38 +676,191 @@ private static IEnumerable<TSource> TakeWhileIterator<TSource>(IEnumerable<TSour
+            {
+                IEnumerator<TSource> sourceEnumerator = _source.GetEnumerator();
+                try
                 {
-                    while (e.MoveNext()) yield return e.Current;
+                    for (int count = _count; count != 0; --count)
+                    {
+                        // It should be harmless to have the recipient call MoveNext again.
+                        // but we guard against strange behaviour from shortcuts in the enumerator
+                        // It should also be harmless to dispose the enumerator here if we've
+                        // exhausted it, but we guard against strange behaviour in either the
+                        // enumerator or the caller.
+                        if (!sourceEnumerator.MoveNext()) return new DeadEnumeratorWrapper<TSource>(sourceEnumerator);
+                    }
+                    return sourceEnumerator;
+                }
+                catch (Exception ex)
+                {
+                    // Force the exception to happen on a MovedNext from the caller, rather than
+                    // here.
+                    return new DeadEnumeratorWrapper<TSource>(sourceEnumerator, ExceptionDispatchInfo.Capture(ex));
                 }
             }
+
+            IEnumerator IEnumerable.GetEnumerator()
+            {
+                return GetEnumerator();
+            }
         }
 
         public static IEnumerable<TSource> SkipWhile<TSource>(this IEnumerable<TSource> source, Func<TSource, bool> predicate)
         {
             if (source == null) throw Error.ArgumentNull("source");
             if (predicate == null) throw Error.ArgumentNull("predicate");
-            return SkipWhileIterator<TSource>(source, predicate);
+            return new SkipWhileIterator<TSource>(source, predicate);
         }
 
-        private static IEnumerable<TSource> SkipWhileIterator<TSource>(IEnumerable<TSource> source, Func<TSource, bool> predicate)
+        private sealed class SkipWhileIterator<TSource> : IEnumerable<TSource>, IEnumerator<TSource>
         {
-            bool yielding = false;
-            foreach (TSource element in source)
+            private readonly IEnumerable<TSource> _source;
+            private readonly Func<TSource, bool> _predicate;
+            private readonly int _threadId = Environment.CurrentManagedThreadId;
+            private IEnumerator<TSource> _enumerator;
+            private int _state;
+            private TSource _current;
+
+            internal SkipWhileIterator(IEnumerable<TSource> source, Func<TSource, bool> predicate)
+            {
+                _source = source;
+                _predicate = predicate;
+            }
+
+            public IEnumerator<TSource> GetEnumerator()
+            {
+                // Don't short-circuit. These are almost always both true, so don't "save"
+                // time by branching after a very likely case.
```

Comment:
I've come across `CurrentManagedThreadId` being expensive enough to worry about in efficiency pushes when doing some highly concurrent code in the past, but was assuming state not being zero would be rare. Since you report that not being the case I'll change this when I'm next not only on a phone.